common/thorhelper/thorxmlread.cpp

```diff
@@ -1653,9 +1654,22 @@ class CXMLParse : public CInterface, implements IXMLParse
         void setMarkingStream(CMarkReadBase &_marking) { marking.set(&_marking); }
         CXPath &queryXPath() { return xpath; }
 
-// IPTreeMaker
+        bool checkSkipJsonNode(const char *tag)
+        {
+            if (!isJson || stack.ordinality()) //json root level only
+                return false;
+            if (streq(tag, "__array__")) //xpath starts after root array
+                return true;
+            if (streq(tag, "__object__") && xpath.queryDepth()) //empty xpath matches start object, otherwise skip, xpath starts immediately after
+                return true;
+            return false;
+        }
+
+        // IPTreeMaker
         virtual void beginNode(const char *tag, offset_t startOffset)
         {
+            if (checkSkipJsonNode(tag))
+                return;
```

Comment:
I think this would look cleaner (and slightly more effieicnt) if you turned CXMLMaker into CMakerBase and derived a CXMLMaker and a CJSonMaker, i.e. where just the CJsonMaker overloaded beginNode/newAttribute/endNode called these methods.